src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -0,0 +1,378 @@
+
+  def genotypeType: Type = metadata.genotypeSignature
+
+  def typ = TStruct(
+    "v" -> TVariant,
+    "va" -> vaType,
+    "s" -> TString,
+    "sa" -> saType,
+    "g" -> genotypeType)
+
+  def sampleEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "g" -> (3, genotypeType),
+      "v" -> (4, TVariant),
+      "va" -> (5, vaType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+
+  def variantEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "g" -> (3, genotypeType),
+      "s" -> (4, TString),
+      "sa" -> (5, saType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+}
+
+object NewAST {
+  def genericRewriteTopDown(ast: NewAST, rule: PartialFunction[NewAST, NewAST]): NewAST = {
+    def rewrite(ast: NewAST): NewAST = {
+      rule.lift(ast) match {
+        case Some(newAST) if newAST ne ast =>
+          rewrite(newAST)
+        case None =>
+          val newChildren = ast.children.map(rewrite)
+          if ((ast.children, newChildren).zipped.forall(_ eq _))
+            ast
+          else
+            ast.copy(newChildren)
+      }
+    }
+
+    rewrite(ast)
+  }
+
+  def rewriteTopDown[T](ast: MatrixAST[T], rule: PartialFunction[NewAST, NewAST]): MatrixAST[T] =
+    genericRewriteTopDown(ast, rule).asInstanceOf[MatrixAST[T]]
+
+  def rewriteTopDown[T](ast: KeyTableAST, rule: PartialFunction[NewAST, NewAST]): KeyTableAST =
+    genericRewriteTopDown(ast, rule).asInstanceOf[KeyTableAST]
+
+  def genericRewriteBottomUp(ast: NewAST, rule: PartialFunction[NewAST, NewAST]): NewAST = {
+    def rewrite(ast: NewAST): NewAST = {
+      val newChildren = ast.children.map(rewrite)
+
+      // only recons if necessary
+      val rewritten =
+        if ((ast.children, newChildren).zipped.forall(_ eq _))
+          ast
+        else
+          ast.copy(newChildren)
+
+      rule.lift(rewritten) match {
+        case Some(newAST) if newAST ne rewritten =>
+          rewrite(newAST)
+        case None =>
+          rewritten
+      }
+    }
+
+    rewrite(ast)
+  }
+  
+  def rewriteBottomUp[T](ast: MatrixAST[T], rule: PartialFunction[NewAST, NewAST]): MatrixAST[T] =
+    genericRewriteBottomUp(ast, rule).asInstanceOf[MatrixAST[T]]
+
+  def rewriteBottomUp[T](ast: KeyTableAST, rule: PartialFunction[NewAST, NewAST]): KeyTableAST =
+    genericRewriteBottomUp(ast, rule).asInstanceOf[KeyTableAST]
+}
+
+abstract class NewAST {
+  def children: IndexedSeq[NewAST]
+
+  def copy(newChildren: IndexedSeq[NewAST]): NewAST
+
+  def mapChildren(f: (NewAST) => NewAST): NewAST = {
```

Comment:
Man, you really go down the types rabbit hole.  I don't understand this.

In general, NewAST[T] is not enough, since AST nodes can have heterogenous children.  We already have that: FilterSamples has a AST (predicate) and child (MatrixAST).

It seems unnatural to me to think of NewAST (or a subclass of NewAST) as just a container for children.  If anything, it's a tree, in which case shouldn't map (1) map over just the children, but all the sub-nodes, and (2) map over the non-child per-node data (e.g. real path, literal value, etc.)?  The second seems like a sound objection independent of the tree-ness.

So mapChildren seems like a real clarification to me.  That said, I don't actually care so much.